keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -244,7 +245,206 @@ There are likely others.
+been done for any of the proposals.
+
+The CoDel technique is described as parameterless but has two magic
+numbers: an "initial interval" of 100 ms and a "target" of 5 ms.  The
+initial interval is set based on round trip times in the Internet, and
+the target is set based on a desired limit on the latency at each hop.
+What are the analogous numbers for our scenario?  We do not have large
+numbers of hops; typically at most two (client to main apiserver and
+then main apiserver to aggregated apiserver).  What is analogous to
+network round trip time?  We have a latency goal of 1 second, and a
+request service time limit of 1 minute.  If we take 1 second as the
+initial interval then note that the maximum service time is much
+larger than the initial interval; by contrast, in networking, the
+maximum service time (i.e., packet length / link speed) is much
+smaller than the initial interval.  Even if we take 1 minute as our
+initial interval, we still do not have the sort of relationship that
+obtains in networking.  Note that in order to get good statistics on a
+queue --- which is needed by the CoDel technique --- there have to be
+many requests served during an interval.  Because of this mismatch,
+and because equivalence of context has not been established, we are
+not agreed that the CoDel technique can be used.
+
+Note that the resource limit being applied is a distinct concept from
+the fairness criteria.  For example, in CPU scheduling there may be 4
+CPUs and 50 threads being scheduled onto those CPUs; we do not suppose
+the goal is to have each thread to be using 0.08 CPUs at each instant;
+a thread uses eitehr 0 or 1 CPUs at a given instant.  Similarly, in
+networking, a router may multiplex a thousand flows onto one link; the
+goal is not to have each flow use 1/1000th of the link at each
+instant; a packet uses 0 links while queued and 1 link while being
+transmitted.  Each CPU or link is used for just one thing at a time;
+this is the resource limit.  The fairness goal is about utilization
+observed over time.  So it is in our scenario too.  For example, we
+may have 5000 flows of requests and a concurrency limit of 600
+requests at any one time.  That does not mean that our goal is for
+each flow to have 0.12 requests running at each instant.  Our goal is
+to limit the number of running requests to 600 at each instant and
+provide some fairness in utilization averaged over time.
+
+That average over time must not be over too much or too little time.
+It would not make sense to average over all past time; that would
+allow a flow to build up a huge amount of credit, enabling it to crowd
+out other flows.  It also does not make sense for the average to cover
+a small amount of time.  Because serving requests, like transmitting
+packets, is lumpy we must average over many service times.  Approaches
+to this include: using a sequence of intervals, using a sliding
+window, and using an exponential decay.
```

Comment:
Something to keep in mind is that we don't want this to be computationally expensive (the overhead we will introduce should be as small as possible). Maybe that argument isn't extremely important in this context, but it potentially may be.

My personal preference would be "sliding window", because it should be relatively cheap and is easy to understand by people.

Reply:
Fortunately, we do not have to explicitly compute averages.  There are scheduling algorithms that achieve fairness without computing the averages.  My proposal is to use a particular one of those.

Reply:
Got it - thanks.